lib/parsers/aac_parser/adts_header_info.rb

```diff
@@ -0,0 +1,134 @@
+  # Will return nil if the header does not match the ADTS specifications.
+  def self.parse_adts_header(header_bits)
+    result = FormatParser::AdtsHeaderInfo.new
+
+    AAC_ADTS_HEADER_BITS_CHUNK_SIZES.each do |letter_size|
+      letter = letter_size[0]
+      chunk_size = letter_size[1]
+      chunk = header_bits.shift(chunk_size)
+      decimal_number = convert_binary_to_decimal(chunk)
+
+      # Skipping data represented by the letters G, K, L, Q, as we are not interested in those values.
+      case letter
+      when 'A'
+        # Syncword, all bits must be set to 1
+        return nil unless chunk.all? { |bit| bit == '1' }
+      when 'B'
+        # MPEG Version, set to 0 for MPEG-4 and 1 for MPEG-2
+        result.mpeg_version = decimal_number
+      when 'C'
+        # Layer, always set to 0
+        return nil unless decimal_number == 0
+      when 'D'
+        # Protection absence, set to 1 if there is no CRC and 0 if there is CRC
+        result.protection_absence = decimal_number == 1
+      when 'E'
+        # AAC Profile
+        return nil unless MPEG4_AUDIO_OBJECT_TYPE_RANGE.include?(decimal_number + 1)
+        result.profile = decimal_number
+      when 'F'
+        # MPEG-4 Sampling Frequency Index (15 is forbidden)
+        return nil unless MPEG4_AUDIO_SAMPLING_FREQUENCY_RANGE.include?(decimal_number)
+        result.mpeg4_sampling_frequency_index = decimal_number
+      when 'H'
+        # MPEG-4 Channel Configuration (in the case of 0, the channel configuration is sent via an in-band PCE (Program Config Element))
+        result.mpeg4_channel_config = decimal_number
+      when 'I'
+        # Originality, set to 1 to signal originality of the audio and 0 otherwise
+        result.originality = decimal_number == 1
+      when 'J'
+        # Home, set to 1 to signal home usage of the audio and 0 otherwise
+        result.home_usage = decimal_number == 1
+      when 'M'
+        # Frame length, length of the ADTS frame including headers and CRC check (protectionabsent == 1? 7: 9)
+        # We expect this to be higher than the header length, but we won't impose any other restrictions
+        header_length = result.protection_absence ? 7 : 9
+        return nil unless decimal_number > header_length
+        result.frame_length = decimal_number
+      when 'O'
+        # Buffer fullness, states the bit-reservoir per frame.
+        # It is merely an informative field with no clear use case defined in the specification.
+        result.buffer_fullness = decimal_number
+      when 'P'
+        # Number of AAC frames (RDBs (Raw Data Blocks)) in ADTS frame minus 1. For maximum compatibility always use one AAC frame per ADTS frame.
+        result.aac_frames_per_adts_frame = decimal_number + 1
+      end
+    end
+
+    result
+  end
+
+  # Converts a binary number given as a array of characters representing bits, into a decimal number.
+  def self.convert_binary_to_decimal(binary_number)
```

Comment:
Could this be make use of `.to_i 2` instead?

Reply:
Done!